### ISSUE TITLE
Security Issue in OpenID Connect Implementation

### ISSUE THREAD

**Issue by tamjidrahat** [oracle/cordova-plugin-oracle-idm-auth#11]
**Security Impact:** Bypass Signature Verification
In the implementation of OpenID Connect Authentication Service (OpenIDConnect10AuthenticationService.java), the library skips the signature verification and successfully authenticates the users when JWK response is not successful. This allows an attacker to simply use a malformed JWK Uri to bypass the signature verification process. This also violates the OpenID Connect specification which requires verifying the signature for ID token before acting on the access token response.

**Relevant file:** https://github.com/oracle/cordova-plugin-oracle-idm-auth/blob/master/src/android/sdk/oracle/idm/mobile/auth/OpenIDConnect10AuthenticationService.java#L113

**How To Fix:** Instead of skipping the signature verification and calling "_onOpenIDSuccess()_" on Line 113 (OpenIDConnect10AuthenticationService.java), the implementation can simply throw _OMMobileSecurityException_ when the value of _verify_ is false or JWKs response is unsuccessful.

**Comment by manish2788** [comment]
Hi Tamjid,
Do you have a sample cordova hybrid mobile app which demonstrates this issue?

**Comment by tamjidrahat** [comment]
I don't have a sample app at this moment.

**Comment by manish2788** [comment]
It’s not actually an issue. According to the spec (https://openid.net/specs/openid-connect-core-1_0.html), there are three different authentication flows available:

1. Authentication using the Authorization Code Flow
2. Authentication using the Implicit Flow
3. Authentication using the Hybrid Flow

Our implementation is using method #1  -- Authentication using the Authorization Code Flow, implemented here: 
https://github.com/oracle/cordova-plugin-oracle-idm-auth/blob/7977c2f1250295b653495c68aae7e0a84b71022c/src/android/sdk/oracle/idm/mobile/auth/OAuthAuthorizationCodeService.java#L254

The spec outlines validating the ID token here: https://openid.net/specs/openid-connect-core-1_0.html#IDTokenValidation, and step 6 is what is pertinent to this issue:
 
“6. If the ID Token is received via direct communication between the Client and the Token Endpoint (which it is in this flow), the TLS server validation MAY be used to validate the issuer in place of checking the token signature. The Client MUST validate the signature of all other ID Tokens according to [JWS](https://openid.net/specs/openid-connect-core-1_0.html#JWS) [JWS] using the algorithm specified in the JWT alg Header Parameter. The Client MUST use the keys provided by the Issuer.”
 
Since we are retrieving the ID token via direct communication between the Client and Token Endpoint (because we are using the Authorization Code Flow), validating the token’s signature isn’t required (because the endpoint has already been deemed trustworthy via TLS). However, in our implementation, we are checking the signature anyway, whenever possible. Whether the validation is possible is determined by whether we can get the signing cert or not which is determined in this code:
 
```
OMHTTPResponse response = getSigningCertForIDCS(url, accessToken);
boolean verify = false;
if (response != null) {
    int responseCode = response.getResponseCode();
    OMLog.debug(TAG, "Response Code: " + responseCode);
    if (responseCode == HttpURLConnection.HTTP_OK) {
        jwksResponse = response.getResponseStringOnSuccess();
        verify = true;
    } else {
        jwksResponse = response.getResponseStringOnFailure();
        verify = false;//no verifying required
    }
}
```
 
In other words, the “verify” variable represents whether signature verification is possible, not whether it is required (despite what the comment says). Which means that we are doubling up on verification whenever we can, and at no point are we neglecting verification altogether.
